implementation/src/main/java/io/smallrye/jwt/auth/principal/KeyLocationResolver.java

```diff
@@ -52,185 +52,226 @@
-            // httpsJwks may refresh itself even if the jwksRefreshInterval is set to 0 as it checks HTTPS Cache-Control header
-            // so the public key will have to be created per request to ensure the rotation (if any) is effective.
-            verificationKey = publicKey;
+            if (jwk != null) {
+                return PublicJsonWebKey.Factory.newPublicJwk(jwk.toString()).getPublicKey();
+            }
+        } catch (Exception e) {
+            LOGGER.debug("Failed to read location as JWK(S)", e);
         }
 
-        return publicKey;
+        return null;
     }
 
-    PublicKey getHttpJwk(HttpsJwks httpsJwks, String kid) throws JoseException, IOException {
-        List<JsonWebKey> keys = httpsJwks.getJsonWebKeys();
+    private String getKid(JsonWebSignature jws) throws UnresolvableKeyException {
+        if (jws.getHeaders() == null)
+            throw new UnresolvableKeyException("Invalid token 'kid' header");
+
+        String kid = jws.getHeaders().getStringHeaderValue(JsonWebKey.KEY_ID_PARAMETER);
         if (kid != null) {
-            for (JsonWebKey currentJwk : keys) {
-                if (kid.equals(currentJwk.getKeyId())) {
-                    return PublicJsonWebKey.class.cast(currentJwk).getPublicKey();
-                }
+            if (authContextInfo.getTokenKeyId() != null && !kid.equals(authContextInfo.getTokenKeyId())) {
+                LOGGER.debugf("Invalid token 'kid' header: %s, expected: %s", kid, authContextInfo.getTokenKeyId());
+                throw new UnresolvableKeyException("Invalid token 'kid' header");
             }
-        } else if (keys.size() == 1) {
-            return PublicJsonWebKey.class.cast(keys.get(0)).getPublicKey();
+        } else {
+            kid = authContextInfo.getTokenKeyId();
         }
-
-        return null;
+        return kid;
     }
 
-    PublicKey getContentJwk(String content, String kid) throws JoseException {
-        JsonObject jwk = null;
+    private static final class HttpsJwksSupplier implements Supplier<HttpsJwks> {
+        private final JWTAuthContextInfo authContextInfo;
 
-        JsonObject jwks;
-        try (JsonReader reader = Json.createReader(new StringReader(content))) {
-            jwks = reader.readObject();
+        public HttpsJwksSupplier(JWTAuthContextInfo authContextInfo) {
+            this.authContextInfo = authContextInfo;
         }
-        JsonArray keys = jwks.getJsonArray(JsonWebKeySet.JWK_SET_MEMBER_NAME);
 
-        if (keys != null) {
-            if (kid != null) {
-                for (int i = 0; i < keys.size(); i++) {
-                    JsonObject currentJwk = keys.getJsonObject(i);
-                    if (kid.equals(currentJwk.getString(JsonWebKey.KEY_ID_PARAMETER, null))) {
-                        jwk = currentJwk;
-                        break;
-                    }
-                }
-            } else if (keys.size() == 1) {
-                jwk = keys.getJsonObject(0);
+        @Override
```

Comment:
Please, as I said, let me see the original loadContent with some simple cleanup if needed and I will be happy :-)

Reply:
loadContent is dangerous, it can be called not only from the constructor and it changes state of the object